### PR TITLE
[JenkinsDownloadArtifacts] Supporting download to a relative path

### PR DESCRIPTION
**Pull request by ds-ms**

### CODE REVIEW

File: Tasks/Common/azure-blobstorage-artifactProvider/azureBlobStorageProvider.ts [hunk 1]
```diff
@@ -61,7 +66,12 @@ export class AzureBlobProvider implements models.IArtifactProvider {
 
     public getArtifactItem(artifactItem: models.ArtifactItem): Promise<NodeJS.ReadableStream> {
```
Comment (sachinma) [inline Tasks/Common/azure-blobstorage-artifactProvider/azureBlobStorageProvider.ts]: Shouldn't getArtifactItems also be changed to add the prefix

File: Tasks/Common/azure-blobstorage-artifactProvider/blobservice.ts [hunk 2]
```diff
@@ -39,7 +39,7 @@ export class BlobService {
 
     public async downloadBlobs(destination: string, container: string, prefixFolderPath?: string, itemPattern?: string): Promise<void> {
         var fileProvider = new artifactProviders.FilesystemProvider(destination);
-        var azureProvider = new azureBlobProvider.AzureBlobProvider(this._storageAccountName, container, this._storageAccessKey, prefixFolderPath, this._host);
+        var azureProvider = new azureBlobProvider.AzureBlobProvider(this._storageAccountName, container, this._storageAccessKey, prefixFolderPath, this._host, true);
```
Comment (sachinma) [inline Tasks/Common/azure-blobstorage-artifactProvider/blobservice.ts]: shouldn't this be false for jenkins

File: Tasks/Common/azure-blobstorage-artifactProvider/azureBlobStorageProvider.ts (hunk 1 above)
Comment (ds-ms) [reply Tasks/Common/azure-blobstorage-artifactProvider/azureBlobStorageProvider.ts]: getArtifactItems removes the prefix, and getArtifactItem appends it

File: Tasks/Common/azure-blobstorage-artifactProvider/blobservice.ts (hunk 2 above)
Comment (ds-ms) [reply Tasks/Common/azure-blobstorage-artifactProvider/blobservice.ts]: No, it should be true for jenkins

File: Tasks/Common/azure-blobstorage-artifactProvider/Strings/resources.resjson/en-US/resources.resjson [hunk 3]
```diff
@@ -9,5 +8,7 @@
   "loc.messages.FailedToListItemInsideContainer": "Failed to list items inside container: {0}. Error: {1}",
   "loc.messages.SuccessFullyFetchedItemList": "Successfully fetcted list of items",
   "loc.messages.UnableToFetchItem": "Unable to fetch item: {0}. Error: {1}",
-  "loc.messages.UploadingItem": "Uploading {0}."
+  "loc.messages.UploadingItem": "Uploading {0}.",
+  "loc.messages.ContinuationTokenExistsFetchingRemainingFiles": "Continuation token exists, trying to fetch the list of remaining files.",
+  "loc.messages.GetArtifactItemNotSupported": "Get artifact item not supported, invalid code path"
```
Comment (hiyadav) [inline Tasks/Common/azure-blobstorage-artifactProvider/Strings/resources.resjson/en-US/resources.resjson]: should this be GetArtifactItems ?

File: Tasks/Common/azure-blobstorage-artifactProvider/azureBlobStorageProvider.ts [hunk 4]
```diff
@@ -56,12 +61,17 @@ export class AzureBlobProvider implements models.IArtifactProvider {
     }
 
     public getArtifactItems(artifactItem: models.ArtifactItem): Promise<models.ArtifactItem[]> {
-        return this._getItems(this._container, artifactItem.path);
+        throw new Error(tl.loc("GetArtifactItemNotSupported"));
```
Comment (hiyadav) [inline Tasks/Common/azure-blobstorage-artifactProvider/azureBlobStorageProvider.ts]: also update here to GetArtifactItem**s**

File: Tasks/Common/azure-blobstorage-artifactProvider/blobservice.ts [hunk 5]
```diff
@@ -37,11 +37,12 @@ export class BlobService {
         return uploadedUrls;
     }
 
-    public async downloadBlobs(destination: string, container: string, prefixFolderPath?: string, itemPattern?: string): Promise<void> {
+    public async downloadBlobs(destination: string, container: string, prefixFolderPath?: string, itemPattern?: string, downloadRelativeToPrefixPath?: boolean): Promise<void> {
         var fileProvider = new artifactProviders.FilesystemProvider(destination);
-        var azureProvider = new azureBlobProvider.AzureBlobProvider(this._storageAccountName, container, this._storageAccessKey, prefixFolderPath, this._host);
+        var azureProvider = new azureBlobProvider.AzureBlobProvider(this._storageAccountName, container, this._storageAccessKey, prefixFolderPath, this._host, true);
```
Comment (hiyadav) [inline Tasks/Common/azure-blobstorage-artifactProvider/blobservice.ts]: we should rather pass downloadRelativeToPrefixPath

File: Tasks/Common/azure-blobstorage-artifactProvider/module.json [hunk 6]
```diff
@@ -10,6 +9,8 @@
     "FailedToListItemInsideContainer": "Failed to list items inside container: {0}. Error: {1}",
     "SuccessFullyFetchedItemList": "Successfully fetcted list of items",
     "UnableToFetchItem": "Unable to fetch item: {0}. Error: {1}",
-    "UploadingItem": "Uploading {0}."
+    "UploadingItem": "Uploading {0}.",
+    "ContinuationTokenExistsFetchingRemainingFiles": "Continuation token exists, trying to fetch the list of remaining files.",
+    "GetArtifactItemNotSupported": "Get artifact item not supported, invalid code path"
```
Comment (hiyadav) [inline Tasks/Common/azure-blobstorage-artifactProvider/module.json]: plural

File: Tasks/JenkinsDownloadArtifacts/task.json [hunk 7]
```diff
@@ -345,6 +344,8 @@
         "UploadingItem": "Uploading {0}.",
         "FoundJobType": "Found job type %s",
         "CannotFindJobType": "Could not detect job type",
-        "InvalidJobName": "Invalid job name %s"
+        "InvalidJobName": "Invalid job name %s",
+        "ContinuationTokenExistsFetchingRemainingFiles": "Continuation token exists, trying to fetch the list of remaining files.",
+        "GetArtifactItemNotSupported": "Get artifact item not supported, invalid code path"
```
Comment (hiyadav) [inline Tasks/JenkinsDownloadArtifacts/task.json]: here as well

File: Tasks/Common/azure-blobstorage-artifactProvider/blobservice.ts (hunk 5 above)
Comment (sachinma) [inline Tasks/Common/azure-blobstorage-artifactProvider/blobservice.ts]: use the function coming to this function

File: Tasks/Common/azure-blobstorage-artifactProvider/blobservice.ts [hunk 8]
```diff
@@ -37,11 +37,12 @@ export class BlobService {
         return uploadedUrls;
     }
 
-    public async downloadBlobs(destination: string, container: string, prefixFolderPath?: string, itemPattern?: string): Promise<void> {
+    public async downloadBlobs(destination: string, container: string, prefixFolderPath?: string, itemPattern?: string, downloadRelativeToPrefixPath?: boolean): Promise<void> {
```
Comment (sachinma) [inline Tasks/Common/azure-blobstorage-artifactProvider/blobservice.ts]: addPrefixToDownloadedItems

File: Tasks/Common/azure-blobstorage-artifactProvider/azureBlobStorageProvider.ts [hunk 9]
```diff
@@ -92,18 +102,35 @@ export class AzureBlobProvider implements models.IArtifactProvider {
     private _getItems(container: string, parentRelativePath?: string): Promise<models.ArtifactItem[]> {
         var promise = new Promise<models.ArtifactItem[]>((resolve, reject) => {
             var items: models.ArtifactItem[] = [];
-
-            this._blobSvc.listBlobsSegmentedWithPrefix(container, parentRelativePath, null, (error, result) => {
+            this._blobSvc.listBlobsSegmentedWithPrefix(container, parentRelativePath, null, async (error, result) => {
```
Comment (sachinma) [inline Tasks/Common/azure-blobstorage-artifactProvider/azureBlobStorageProvider.ts]: You can simplify this code and just keep one do-while loop

**Pull request merged**